File: recompiler-rs/scripts/dump_decode.py
#!/usr/bin/env python3
"""Differential-oracle shim: dump the Python `decode_function` output for every
cfg entry to JSON, so the Rust decoder port can be diffed against it.

The env is CONTROLLED and explicit so both sides decode under identical inputs:
the aggregate data_regions / reloc_regions / indirect_dispatch / inline-dispatch
-loop / callee_inline_skip drawn from the cfgs, and NO dispatch_helpers /
callee_exit_mx (those come from later discovery passes; layered in once the core
walk matches). Output is one JSON object per entry, keyed by "bank:pc:m:x".

Usage:
  dump_decode.py --rom sf.sfc --cfg-dir recomp --out golden/decode.json
"""
import argparse
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
# HERE = .../recompiler-rs/scripts → repo root is two levels up.
SNESRECOMP = os.path.dirname(os.path.dirname(HERE))
sys.path.insert(0, os.path.join(SNESRECOMP, "recompiler"))
sys.path.insert(0, os.path.join(SNESRECOMP, "recompiler", "v2"))

from snes65816 import load_rom, set_active_reloc_regions  # noqa: E402
from v2.cfg_loader import load_bank_cfg  # noqa: E402
from v2.decoder import decode_function  # noqa: E402
import glob  # noqa: E402


def key_str(k):
    ps = ";".join(f"{m}{x}" for (m, x) in k.p_stack)
    return f"{k.pc:06X}:{k.m}:{k.x}:[{ps}]"


def insn_obj(di):
    ins = di.insn
    return {
        "addr": ins.addr,
        "op": ins.opcode,
        "mnem": ins.mnem,
        "mode": ins.mode,
        "operand": ins.operand,
        "length": ins.length,
        "m": ins.m_flag,
        "x": ins.x_flag,
        "succ": sorted(key_str(s) for s in di.successors),
    }


def graph_obj(g):
    return {
        "entry": key_str(g.entry),
        "insns": {key_str(k): insn_obj(di) for k, di in g.insns.items()},
        "unresolved": sorted(u.site_pc24 for u in g.unresolved_indirects),
        "suppressed_calls": sorted(s.site_pc24 for s in g.suppressed_indirect_calls),
        "dispatch_suppressed": sorted(
            (d.site_pc24, d.target_pc24) for d in g.dispatch_targets_suppressed
        ),
    }


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--rom", required=True)
    ap.add_argument("--cfg-dir", required=True)
    ap.add_argument("--out", required=True)
    args = ap.parse_args()

    rom = load_rom(args.rom)
    cfgs = []
    for p in sorted(glob.glob(os.path.join(args.cfg_dir, "bank*.cfg"))):
        try:
            cfgs.append((p, load_bank_cfg(p)))
        except Exception as e:
            print(f"  parse-fail {p}: {e}", file=sys.stderr)

    # Aggregate the controlled env across all cfgs.
    data_regions = []
    reloc_regions = []
    callee_inline_skip = {}
    inline_loop_pcs = set()
    indirect_dispatch = {}  # site_pc24 -> {count, idx_reg, table_bases}
    for _p, c in cfgs:
        data_regions.extend(c.data_regions)
        reloc_regions.extend(c.reloc_regions)
        for e in c.entries:
            if getattr(e, "inline_skip", None) is not None:
                callee_inline_skip[(c.bank << 16) | (e.start & 0xFFFF)] = e.inline_skip
        for pc16 in c.inline_dispatch_loops:
            inline_loop_pcs.add((c.bank << 16) | pc16)
        for d in c.indirect_dispatch:
            site = (c.bank << 16) | d["site_pc16"]
            indirect_dispatch[site] = {
                "count": d["count"],
                "idx_reg": d["idx_reg"],
                "table_bases": list(d["table_bases"]),
            }

    set_active_reloc_regions(reloc_regions or None)

    out = {}
    n_ok = n_err = 0
    for _p, c in cfgs:
        for e in c.entries:
            try:
                g = decode_function(
                    rom, c.bank, e.start,
                    entry_m=e.entry_m, entry_x=e.entry_x, end=e.end,
                    data_regions=data_regions or None,
                    reloc_regions=reloc_regions or None,
                    callee_inline_skip=callee_inline_skip or None,
                    indirect_dispatch=indirect_dispatch or None,
                    inline_dispatch_loop_pcs=inline_loop_pcs or None,
                )
            except Exception as ex:
                n_err += 1
                out[f"{c.bank:02X}:{e.start:04X}:{e.entry_m}:{e.entry_x}"] = {
                    "error": f"{type(ex).__name__}: {ex}"
                }
                continue
            n_ok += 1
            out[f"{c.bank:02X}:{e.start:04X}:{e.entry_m}:{e.entry_x}"] = graph_obj(g)

    with open(args.out, "w") as f:
        json.dump(out, f, sort_keys=True, separators=(",", ":"))
    print(f"dumped {n_ok} graphs ({n_err} errors) -> {args.out}")


if __name__ == "__main__":
    main()

File: recompiler-rs/src/bin/dump_decode.rs
//! `dump-decode` — Rust side of the differential oracle.
//!
//! Loads the ROM + cfg dir, builds the SAME controlled `DecodeEnv` as
//! `scripts/dump_decode.py` (aggregate data_regions / reloc_regions /
//! callee_inline_skip / indirect_dispatch / inline_dispatch_loop_pcs; everything
//! else None), decodes every cfg entry, and emits JSON in the identical shape so
//! `scripts/diff_decode.py` can compare against `golden/decode.json`.

use std::collections::{BTreeMap, BTreeSet, HashMap};
use std::panic::{catch_unwind, AssertUnwindSafe};
use std::path::PathBuf;

use serde_json::{json, Map, Value};

use snesrecomp_regen::cfg::load_bank_cfg;
use snesrecomp_regen::decoder::{
    decode_function, DecodeEnv, DecodeKey, DecodedInsn, FunctionDecodeGraph, IndirectDispatchSite,
};
use snesrecomp_regen::rom::{load_rom, RelocRegion};

fn arg_value(args: &[String], flag: &str) -> Option<String> {
    args.iter().position(|a| a == flag).and_then(|i| args.get(i + 1).cloned())
}

fn key_str(k: &DecodeKey) -> String {
    let ps: Vec<String> = k.p_stack.iter().map(|(m, x)| format!("{m}{x}")).collect();
    format!("{:06X}:{}:{}:[{}]", k.pc, k.m, k.x, ps.join(";"))
}

fn insn_obj(di: &DecodedInsn) -> Value {
    let ins = &di.insn;
    let mut succ: Vec<String> = di.successors.iter().map(key_str).collect();
    succ.sort();
    json!({
        "addr": ins.addr,
        "op": ins.opcode,
        "mnem": ins.mnem,
        "mode": ins.mode.index(),
        "operand": ins.operand,
        "length": ins.length,
        "m": ins.m_flag,
        "x": ins.x_flag,
        "succ": succ,
    })
}

fn graph_obj(g: &FunctionDecodeGraph) -> Value {
    let mut insns = Map::new();
    for di in g.insns() {
        insns.insert(key_str(&di.key), insn_obj(di));
    }
    let mut unresolved: Vec<u32> = g.unresolved_indirects.iter().map(|u| u.site_pc24).collect();
    unresolved.sort();
    let mut suppressed: Vec<u32> =
        g.suppressed_indirect_calls.iter().map(|s| s.site_pc24).collect();
    suppressed.sort();
    let mut dsupp: Vec<(u32, u32)> = g
        .dispatch_targets_suppressed
        .iter()
        .map(|d| (d.site_pc24, d.target_pc24))
        .collect();
    dsupp.sort();
    let dsupp_v: Vec<Value> = dsupp.iter().map(|(a, b)| json!([a, b])).collect();

    json!({
        "entry": g.entry.as_ref().map(key_str).unwrap_or_default(),
        "insns": Value::Object(insns),
        "unresolved": unresolved,
        "suppressed_calls": suppressed,
        "dispatch_suppressed": dsupp_v,
    })
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let rom_path = arg_value(&args, "--rom").expect("--rom required");
    let cfg_dir = arg_value(&args, "--cfg-dir").expect("--cfg-dir required");
    let out_path = arg_value(&args, "--out").expect("--out required");

    let rom = load_rom(&rom_path).expect("load rom");

    // Glob bank*.cfg sorted (matches Python sorted(glob)).
    let mut cfg_paths: Vec<PathBuf> = std::fs::read_dir(&cfg_dir)
        .expect("read cfg dir")
        .filter_map(|e| e.ok().map(|e| e.path()))
        .filter(|p| {
            p.file_name()
                .and_then(|n| n.to_str())
                .map(|n| n.starts_with("bank") && n.ends_with(".cfg"))
                .unwrap_or(false)
        })
        .collect();
    cfg_paths.sort();

    let mut cfgs = Vec::new();
    for p in &cfg_paths {
        match load_bank_cfg(p) {
            Ok(c) => cfgs.push(c),
            Err(e) => eprintln!("  parse-fail {}: {e}", p.display()),
        }
    }

    // Aggregate the controlled env across all cfgs.
    let mut data_regions: Vec<(u32, u32, u32)> = Vec::new();
    let mut reloc_regions: Vec<RelocRegion> = Vec::new();
    let mut callee_inline_skip: HashMap<u32, i32> = HashMap::new();
    let mut inline_loop_pcs: BTreeSet<u32> = BTreeSet::new();
    let mut indirect_dispatch: HashMap<u32, IndirectDispatchSite> = HashMap::new();

    for c in &cfgs {
        let bank = c.bank as u32;
        data_regions.extend(c.data_regions.iter().copied());
        reloc_regions.extend(c.reloc_regions.iter().copied());
        for e in &c.entries {
            if let Some(skip) = e.inline_skip {
                callee_inline_skip.insert((bank << 16) | (e.start & 0xFFFF), skip);
            }
        }
        for &pc16 in &c.inline_dispatch_loops {
            inline_loop_pcs.insert((bank << 16) | pc16);
        }
        for d in &c.indirect_dispatch {
            let site = (bank << 16) | d.site_pc16;
            indirect_dispatch.insert(
                site,
                IndirectDispatchSite {
                    count: d.count,
                    idx_reg: d.idx_reg,
                    table_bases: d.table_bases.clone(),
                },
            );
        }
    }

    let env = DecodeEnv {
        data_regions: Some(&data_regions),
        reloc_regions: Some(&reloc_regions),
        callee_inline_skip: Some(&callee_inline_skip),
        indirect_dispatch: Some(&indirect_dispatch),
        inline_dispatch_loop_pcs: Some(&inline_loop_pcs),
        ..Default::default()
    };

    let mut out: BTreeMap<String, Value> = BTreeMap::new();
    let mut n_ok = 0usize;
    let mut n_err = 0usize;
    for c in &cfgs {
        let bank = c.bank as u32;
        for e in &c.entries {
            let key = format!("{:02X}:{:04X}:{}:{}", c.bank, e.start, e.entry_m, e.entry_x);
            let res = catch_unwind(AssertUnwindSafe(|| {
                decode_function(&rom, bank, e.start, e.entry_m, e.entry_x, e.end, &env)
            }));
            match res {
                Ok(g) => {
                    n_ok += 1;
                    out.insert(key, graph_obj(&g));
                }
                Err(_) => {
                    n_err += 1;
                    out.insert(key, json!({ "error": "panic" }));
                }
            }
        }
    }

    let val = Value::Object(out.into_iter().collect());
    let s = serde_json::to_string(&val).expect("serialize");
    std::fs::write(&out_path, s).expect("write out");
    eprintln!("dumped {n_ok} graphs ({n_err} errors) -> {out_path}");
}
